File: bindings/python/python/auths/_production_client.py
from __future__ import annotations

import asyncio
import base64
import json
import ssl
import urllib.error
import urllib.request
from dataclasses import dataclass
from types import MappingProxyType
from typing import Literal, Mapping, Optional, Protocol, Union, cast, runtime_checkable
from urllib.parse import urlparse

from ._native import (
    decode_production_response_v1,
    encode_production_delegation_v1,
    encode_production_request_v1,
    production_client_contract_version_v1,
)
from .profiles import ProductionProfile

_CONTENT_TYPE = "application/auths+cbor"
_MAX_RESPONSE_BYTES = 1_048_576
_DEFAULT_TIMEOUT_SECONDS = 15.0
_AUTHORITY_TOKEN = object()
_RECEIPT_TOKEN = object()
_REFERENCE_TOKEN = object()

ProductStep = Literal["create", "delegate", "execute", "resume", "verify"]
RetryClass = Literal["never", "backoff", "resume", "reconcile"]


@dataclass(frozen=True)
class ProductionTransportRequest:
    url: str
    body: bytes
    content_type: Literal["application/auths+cbor"]
    timeout_seconds: float


@dataclass(frozen=True)
class ProductionTransportResponse:
    status: int
    content_type: str
    body: bytes


@runtime_checkable
class ProductionTransport(Protocol):
    async def send(
        self, request: ProductionTransportRequest
    ) -> ProductionTransportResponse: ...


class ProductionAuthority:
    __slots__ = ("_bytes",)
    kind: Literal["authority"] = "authority"

    def __init__(self, token: object, value: bytes) -> None:
        if token is not _AUTHORITY_TOKEN or not value:
            raise TypeError("sealed Auths authority")
        self._bytes = bytes(value)

    def __reduce__(self) -> None:
        raise TypeError("Auths authority is opaque")


class ProductionReceipt:
    __slots__ = ("_bytes",)
    kind: Literal["receipt"] = "receipt"

    def __init__(self, token: object, value: bytes) -> None:
        if token is not _RECEIPT_TOKEN or not value:
            raise TypeError("sealed Auths receipt")
        self._bytes = bytes(value)

    def __reduce__(self) -> None:
        raise TypeError("Auths receipt bytes require an explicit disclosure operation")


class ProductionRecoveryReference:
    __slots__ = ("_value",)
    kind: Literal["recovery-reference"] = "recovery-reference"

    def __init__(self, token: object, value: str) -> None:
        if token is not _REFERENCE_TOKEN or not _is_recovery_reference(value):
            raise TypeError("sealed Auths recovery reference")
        self._value = value

    def __reduce__(self) -> None:
        raise TypeError("Auths recovery references are opaque")


@dataclass(frozen=True)
class ProductionDenied:
    kind: Literal["denied"]
    step: ProductStep
    code: str
    retry: Literal["never"]


@dataclass(frozen=True)
class ProductionIndeterminate:
    kind: Literal["indeterminate"]
    step: ProductStep
    code: str
    retry: Literal["backoff", "reconcile"]


@dataclass(frozen=True)
class ProductionRecoverable:
    kind: Literal["recoverable"]
    step: Literal["execute", "resume"]
    code: str
    retry: Literal["resume"]
    reference: ProductionRecoveryReference


@dataclass(frozen=True)
class ProductionCompleted:
    kind: Literal["completed"]
    step: Literal["execute", "resume"]
    value: Optional[bytes]
    receipt: ProductionReceipt


@dataclass(frozen=True)
class ProductionVerified:
    kind: Literal["verified"]
    step: Literal["verify"]
    value: Optional[bytes]


@dataclass(frozen=True)
class ProductionRejected:
    kind: Literal["rejected"]
    step: Literal["verify"]
    code: str
    retry: Literal["never"]


ProductionAuthorityResult = Union[
    ProductionAuthority, ProductionDenied, ProductionIndeterminate
]
ProductionExecutionResult = Union[
    ProductionCompleted,
    ProductionDenied,
    ProductionIndeterminate,
    ProductionRecoverable,
]
ProductionVerificationResult = Union[
    ProductionVerified, ProductionRejected, ProductionIndeterminate
]


class ProductionAuths:
    def __init__(
        self,
        *,
        endpoint: str,
        identity: bytes,
        profile: ProductionProfile,
        transport: Optional[ProductionTransport] = None,
        timeout_seconds: float = _DEFAULT_TIMEOUT_SECONDS,
    ) -> None:
        self._endpoint = _parse_endpoint(endpoint)
        self._identity = _bounded_bytes(identity, 65_536, "identity")
        if type(profile) is not ProductionProfile or profile.id not in _PROFILE_IDS:
            raise TypeError("Auths production profile is unsupported")
        if (
            type(timeout_seconds) not in (int, float)
            or timeout_seconds < 0.1
            or timeout_seconds > 120
        ):
            raise ValueError("Auths production timeout is outside bounds")
        if transport is not None and not isinstance(transport, ProductionTransport):
            raise TypeError("Auths production transport is invalid")
        self._profile = profile
        self._transport = transport or _UrlLibProductionTransport()
        self._timeout_seconds = float(timeout_seconds)

    async def create(self, request: bytes) -> ProductionAuthorityResult:
        projection = await self._call("create", body=request)
        if projection["kind"] == "completed":
            return ProductionAuthority(_AUTHORITY_TOKEN, _required_value(projection))
        return _authority_failure("create", projection)

    async def delegate(
        self,
        authority: ProductionAuthority,
        subject: bytes,
        attenuation: bytes = b"\x80",
    ) -> ProductionAuthorityResult:
        body = bytes(
            encode_production_delegation_v1(
                _bounded_bytes(subject, 65_536, "subject"),
                _bounded_bytes(attenuation, 65_536, "attenuation"),
            )
        )
        projection = await self._call(
            "delegate", authority=_authority_bytes(authority), body=body
        )
        if projection["kind"] == "completed":
            return ProductionAuthority(_AUTHORITY_TOKEN, _required_value(projection))
        return _authority_failure("delegate", projection)

    async def execute(
        self, authority: ProductionAuthority, action: bytes
    ) -> ProductionExecutionResult:
        return _execution_result(
            "execute",
            await self._call(
                "execute", authority=_authority_bytes(authority), body=action
            ),
        )

    async def resume(
        self, reference: ProductionRecoveryReference
    ) -> ProductionExecutionResult:
        if type(reference) is not ProductionRecoveryReference:
            raise TypeError("forged Auths recovery reference")
        return _execution_result(
            "resume",
            await self._call("resume", recovery_reference=reference._value),
        )

    async def verify(
        self, value: Union[ProductionAuthority, ProductionReceipt, bytes]
    ) -> ProductionVerificationResult:
        if type(value) is ProductionAuthority:
            body = value._bytes
        elif type(value) is ProductionReceipt:
            body = value._bytes
        else:
            body = _bounded_bytes(value, _MAX_RESPONSE_BYTES, "verification input")
        projection = await self._call("verify", body=body)
        if projection["kind"] == "verified":
            return ProductionVerified("verified", "verify", _optional_bytes(projection["value"]))
        if projection["kind"] == "rejected":
            return ProductionRejected(
                "rejected", "verify", _required_code(projection), "never"
            )
        if projection["kind"] == "indeterminate":
            return _indeterminate("verify", projection)
        raise TypeError("native response outcome does not match verify")

    async def _call(
        self,
        step: ProductStep,
        *,
        authority: Optional[bytes] = None,
        body: Optional[bytes] = None,
        recovery_reference: Optional[str] = None,
    ) -> Mapping[str, object]:
        request_body = bytes(
            encode_production_request_v1(
                step,
                self._profile.id,
                self._identity,
                authority,
                None if body is None else _bounded_bytes(body, _MAX_RESPONSE_BYTES, "body"),
                recovery_reference,
            )
        )
        request = ProductionTransportRequest(
            self._endpoint + _endpoint_path(step, self._profile.id),
            request_body,
            _CONTENT_TYPE,
            self._timeout_seconds,
        )
        try:
            response = await self._transport.send(request)
        except Exception:
            return MappingProxyType(
                {
                    "contractVersion": 1,
                    "kind": "indeterminate",
                    "code": "core.runtime-unavailable",
                    "retry": "backoff",
                    "recoveryReference": None,
                    "value": None,
                    "receipt": None,
                }
            )
        if (
            not 200 <= response.status < 300
            or response.content_type.split(";", 1)[0].strip().lower() != _CONTENT_TYPE
            or not response.body
            or len(response.body) > _MAX_RESPONSE_BYTES
        ):
            return MappingProxyType(
                {
                    "contractVersion": 1,
                    "kind": "indeterminate",
                    "code": "core.malformed-input",
                    "retry": "backoff",
                    "recoveryReference": None,
                    "value": None,
                    "receipt": None,
                }
            )
        return _projection(decode_production_response_v1(response.body))


def create_auths(
    *,
    endpoint: str,
    identity: bytes,
    profile: ProductionProfile,
    transport: Optional[ProductionTransport] = None,
    timeout_seconds: float = _DEFAULT_TIMEOUT_SECONDS,
) -> ProductionAuths:
    if production_client_contract_version_v1() != 1:
        raise RuntimeError("Auths production client contract mismatch")
    return ProductionAuths(
        endpoint=endpoint,
        identity=identity,
        profile=profile,
        transport=transport,
        timeout_seconds=timeout_seconds,
    )


class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, request: object, *args: object, **kwargs: object) -> None:
        return None


class _UrlLibProductionTransport:
    async def send(
        self, request: ProductionTransportRequest
    ) -> ProductionTransportResponse:
        loop = asyncio.get_running_loop()
        return await loop.run_in_executor(None, _send_sync, request)


def _send_sync(request: ProductionTransportRequest) -> ProductionTransportResponse:
    opener = urllib.request.build_opener(
        urllib.request.HTTPSHandler(context=ssl.create_default_context()), _NoRedirect()
    )
    value = urllib.request.Request(
        request.url,
        data=request.body,
        headers={"Content-Type": request.content_type, "Accept": request.content_type},
        method="POST",
    )
    try:
        response = opener.open(value, timeout=request.timeout_seconds)
    except urllib.error.HTTPError as error:
        response = error
    with response:
        declared = response.headers.get("Content-Length")
        if declared is not None and int(declared) > _MAX_RESPONSE_BYTES:
            raise ValueError("Auths production response is outside bounds")
        body = response.read(_MAX_RESPONSE_BYTES + 1)
        if len(body) > _MAX_RESPONSE_BYTES:
            raise ValueError("Auths production response is outside bounds")
        return ProductionTransportResponse(
            response.status, response.headers.get("Content-Type", ""), body
        )


_PROFILE_IDS = (
    "auths.opentofu.saved-plan-apply/1",
    "auths.postgresql.bounded-update/1",
    "auths.github.issue-address/1",
)


def _parse_endpoint(value: str) -> str:
    if type(value) is not str:
        raise TypeError("Auths production endpoint must be an HTTPS origin")
    parsed = urlparse(value)
    if (
        parsed.scheme != "https"
        or not parsed.hostname
        or parsed.username is not None
        or parsed.password is not None
        or parsed.path not in ("", "/")
        or parsed.params
        or parsed.query
        or parsed.fragment
    ):
        raise ValueError("Auths production endpoint must be an HTTPS origin")
    return value.rstrip("/")


def _endpoint_path(step: ProductStep, profile: str) -> str:
    if step == "create":
        return "/v1/authority/create"
    if step == "delegate":
        return "/v1/authority/delegate"
    if step == "resume":
        return "/v1/workflows/resume"
    if step == "verify":
        return "/v1/authority/verify"
    return {
        _PROFILE_IDS[0]: "/v1/profiles/opentofu/saved-plan-apply/execute",
        _PROFILE_IDS[1]: "/v1/profiles/postgresql/bounded-update/execute",
        _PROFILE_IDS[2]: "/v1/profiles/github/issue-address/execute",
    }[profile]


def _projection(value: str) -> Mapping[str, object]:
    parsed = json.loads(value)
    if (
        type(parsed) is not dict
        or parsed.get("contractVersion") != 1
        or parsed.get("kind")
        not in (
            "completed",
            "denied",
            "indeterminate",
            "recoverable",
            "verified",
            "rejected",
        )
        or parsed.get("retry") not in ("never", "backoff", "resume", "reconcile")
    ):
        raise TypeError("native response projection is invalid")
    return MappingProxyType(parsed)


def _authority_failure(
    step: Literal["create", "delegate"], projection: Mapping[str, object]
) -> Union[ProductionDenied, ProductionIndeterminate]:
    if projection["kind"] == "denied":
        return ProductionDenied("denied", step, _required_code(projection), "never")
    if projection["kind"] == "indeterminate":
        return _indeterminate(step, projection)
    raise TypeError("native response outcome does not match " + step)


def _execution_result(
    step: Literal["execute", "resume"], projection: Mapping[str, object]
) -> ProductionExecutionResult:
    if projection["kind"] == "completed":
        receipt = _optional_bytes(projection["receipt"])
        if receipt is None:
            raise TypeError("native response omitted receipt bytes")
        return ProductionCompleted(
            "completed",
            step,
            _optional_bytes(projection["value"]),
            ProductionReceipt(_RECEIPT_TOKEN, receipt),
        )
    if projection["kind"] == "denied":
        return ProductionDenied("denied", step, _required_code(projection), "never")
    if projection["kind"] == "indeterminate":
        return _indeterminate(step, projection)
    reference = projection.get("recoveryReference")
    if projection["kind"] == "recoverable" and type(reference) is str:
        return ProductionRecoverable(
            "recoverable",
            step,
            _required_code(projection),
            "resume",
            ProductionRecoveryReference(_REFERENCE_TOKEN, reference),
        )
    raise TypeError("native response outcome does not match " + step)


def _indeterminate(
    step: ProductStep, projection: Mapping[str, object]
) -> ProductionIndeterminate:
    retry = projection["retry"]
    if retry not in ("backoff", "reconcile"):
        raise TypeError("native indeterminate result has invalid retry class")
    return ProductionIndeterminate(
        "indeterminate",
        step,
        _required_code(projection),
        cast(Literal["backoff", "reconcile"], retry),
    )


def _required_code(projection: Mapping[str, object]) -> str:
    value = projection.get("code")
    if type(value) is not str:
        raise TypeError("native response omitted stable error code")
    return value


def _required_value(projection: Mapping[str, object]) -> bytes:
    value = _optional_bytes(projection.get("value"))
    if value is None:
        raise TypeError("native response omitted value bytes")
    return value


def _optional_bytes(value: object) -> Optional[bytes]:
    if value is None:
        return None
    if type(value) is not str:
        raise TypeError("native response encoding is invalid")
    try:
        return base64.urlsafe_b64decode(value + "=" * (-len(value) % 4))
    except ValueError:
        raise TypeError("native response encoding is invalid") from None


def _bounded_bytes(value: object, maximum: int, name: str) -> bytes:
    if type(value) is not bytes or not value or len(value) > maximum:
        raise ValueError("Auths " + name + " bytes are outside bounds")
    return value


def _authority_bytes(value: ProductionAuthority) -> bytes:
    if type(value) is not ProductionAuthority:
        raise TypeError("forged Auths authority")
    return value._bytes


def _is_recovery_reference(value: str) -> bool:
    if type(value) is not str or len(value) != 43:
        return False
    try:
        decoded = base64.urlsafe_b64decode(value + "=")
    except ValueError:
        return False
    return len(decoded) == 32 and decoded != bytes(32)


__all__ = [
    "ProductStep",
    "ProductionAuths",
    "ProductionAuthority",
    "ProductionAuthorityResult",
    "ProductionCompleted",
    "ProductionDenied",
    "ProductionExecutionResult",
    "ProductionIndeterminate",
    "ProductionReceipt",
    "ProductionRecoverable",
    "ProductionRecoveryReference",
    "ProductionRejected",
    "ProductionTransport",
    "ProductionTransportRequest",
    "ProductionTransportResponse",
    "ProductionVerificationResult",
    "ProductionVerified",
    "RetryClass",
    "create_auths",
]

File: bindings/typescript/src/production-client.ts
import type { ProductionProfile } from "./profiles.js";
import { loadPackagedWorkflowEngine } from "./verifier/wasm.js";

const CONTENT_TYPE = "application/auths+cbor";
const MAX_RESPONSE_BYTES = 1_048_576;
const DEFAULT_TIMEOUT_MS = 15_000;
const authorityBytes = new WeakMap<ProductionAuthority, Uint8Array>();
const receiptBytes = new WeakMap<ProductionReceipt, Uint8Array>();
const referenceValues = new WeakMap<ProductionRecoveryReference, string>();

export type ProductStep = "create" | "delegate" | "execute" | "resume" | "verify";
export type RetryClass = "never" | "backoff" | "resume" | "reconcile";

export interface ProductionTransportRequest {
  readonly url: URL;
  readonly body: Uint8Array;
  readonly contentType: typeof CONTENT_TYPE;
  readonly timeoutMs: number;
}

export interface ProductionTransportResponse {
  readonly status: number;
  readonly contentType: string;
  readonly body: Uint8Array;
}

export interface ProductionTransport {
  send(request: ProductionTransportRequest): Promise<ProductionTransportResponse>;
}

export interface ProductionAuthsOptions {
  readonly endpoint: string | URL;
  readonly identity: Uint8Array;
  readonly profile: ProductionProfile;
  readonly transport?: ProductionTransport;
  readonly timeoutMs?: number;
}

export interface ProductionAuthority {
  readonly kind: "authority";
  toJSON(): never;
}

class ProductionAuthorityValue implements ProductionAuthority {
  readonly kind = "authority" as const;

  constructor(bytes: Uint8Array) {
    authorityBytes.set(this, bytes.slice());
    Object.freeze(this);
  }

  toJSON(): never {
    throw new TypeError("Auths authority is opaque");
  }
}

export interface ProductionReceipt {
  readonly kind: "receipt";
  toJSON(): never;
}

class ProductionReceiptValue implements ProductionReceipt {
  readonly kind = "receipt" as const;

  constructor(bytes: Uint8Array) {
    receiptBytes.set(this, bytes.slice());
    Object.freeze(this);
  }

  toJSON(): never {
    throw new TypeError("Auths receipt bytes require an explicit disclosure operation");
  }
}

export interface ProductionRecoveryReference {
  readonly kind: "recovery-reference";
  toJSON(): never;
}

class ProductionRecoveryReferenceValue implements ProductionRecoveryReference {
  readonly kind = "recovery-reference" as const;

  constructor(value: string) {
    referenceValues.set(this, value);
    Object.freeze(this);
  }

  toJSON(): never {
    throw new TypeError("Auths recovery references are opaque");
  }
}

export interface ProductionDenied {
  readonly kind: "denied";
  readonly step: ProductStep;
  readonly code: string;
  readonly retry: "never";
}

export interface ProductionIndeterminate {
  readonly kind: "indeterminate";
  readonly step: ProductStep;
  readonly code: string;
  readonly retry: "backoff" | "reconcile";
}

export interface ProductionRecoverable {
  readonly kind: "recoverable";
  readonly step: "execute" | "resume";
  readonly code: string;
  readonly retry: "resume";
  readonly reference: ProductionRecoveryReference;
}

export interface ProductionCompleted {
  readonly kind: "completed";
  readonly step: "execute" | "resume";
  readonly value?: Uint8Array;
  readonly receipt: ProductionReceipt;
}

export interface ProductionVerified {
  readonly kind: "verified";
  readonly step: "verify";
  readonly value?: Uint8Array;
}

export interface ProductionRejected {
  readonly kind: "rejected";
  readonly step: "verify";
  readonly code: string;
  readonly retry: "never";
}

export type ProductionAuthorityResult = ProductionAuthority | ProductionDenied | ProductionIndeterminate;
export type ProductionExecutionResult = ProductionCompleted | ProductionDenied | ProductionIndeterminate | ProductionRecoverable;
export type ProductionVerificationResult = ProductionVerified | ProductionRejected | ProductionIndeterminate;

export interface ProductionAuths {
  create(request: Uint8Array): Promise<ProductionAuthorityResult>;
  delegate(
    authority: ProductionAuthority,
    subject: Uint8Array,
    attenuation?: Uint8Array,
  ): Promise<ProductionAuthorityResult>;
  execute(authority: ProductionAuthority, action: Uint8Array): Promise<ProductionExecutionResult>;
  resume(reference: ProductionRecoveryReference): Promise<ProductionExecutionResult>;
  verify(value: ProductionAuthority | ProductionReceipt | Uint8Array): Promise<ProductionVerificationResult>;
}

interface NativeProductionEngine {
  productionClientContractVersionV1(): number;
  encodeProductionRequestV1(input: Readonly<{
    readonly verb: ProductStep;
    readonly profile: string;
    readonly identity: Uint8Array;
    readonly authority?: Uint8Array;
    readonly body?: Uint8Array;
    readonly recoveryReference?: string;
  }>): Uint8Array;
  decodeProductionResponseV1(input: Uint8Array): string;
  decodeProductionRequestV1(input: Uint8Array): string;
  encodeProductionDelegationV1(subject: Uint8Array, attenuation: Uint8Array): Uint8Array;
}

interface NativeProjection {
  readonly contractVersion: number;
  readonly kind: "completed" | "denied" | "indeterminate" | "recoverable" | "verified" | "rejected";
  readonly code: string | null;
  readonly retry: RetryClass;
  readonly recoveryReference: string | null;
  readonly value: string | null;
  readonly receipt: string | null;
}

class ProductionAuthsClient implements ProductionAuths {
  readonly #endpoint: URL;
  readonly #identity: Uint8Array;
  readonly #profile: ProductionProfile;
  readonly #transport: ProductionTransport;
  readonly #timeoutMs: number;

  constructor(options: ProductionAuthsOptions) {
    this.#endpoint = parseEndpoint(options.endpoint);
    if (!(options.identity instanceof Uint8Array) || options.identity.length === 0 || options.identity.length > 65_536) {
      throw new TypeError("Auths identity bytes are outside production bounds");
    }
    if (!isProductionProfile(options.profile)) throw new TypeError("Auths production profile is unsupported");
    const timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    if (!Number.isSafeInteger(timeoutMs) || timeoutMs < 100 || timeoutMs > 120_000) {
      throw new TypeError("Auths production timeout is outside bounds");
    }
    this.#identity = options.identity.slice();
    this.#profile = options.profile;
    this.#transport = options.transport ?? fetchProductionTransport;
    this.#timeoutMs = timeoutMs;
  }

  async create(request: Uint8Array): Promise<ProductionAuthorityResult> {
    const projection = await this.#call("create", request);
    if (projection.kind === "completed") {
      return productionAuthority(requiredValue(projection));
    }
    return projectAuthorityFailure("create", projection);
  }

  async delegate(
    authority: ProductionAuthority,
    subject: Uint8Array,
    attenuation: Uint8Array = new Uint8Array([0x80]),
  ): Promise<ProductionAuthorityResult> {
    const authorityValue = readAuthority(authority);
    const body = await encodeDelegationInput(subject, attenuation);
    const projection = await this.#call("delegate", body, authorityValue);
    if (projection.kind === "completed") {
      return productionAuthority(requiredValue(projection));
    }
    return projectAuthorityFailure("delegate", projection);
  }

  async execute(authority: ProductionAuthority, action: Uint8Array): Promise<ProductionExecutionResult> {
    return projectExecution("execute", await this.#call("execute", action, readAuthority(authority)));
  }

  async resume(reference: ProductionRecoveryReference): Promise<ProductionExecutionResult> {
    const value = referenceValues.get(reference);
    if (value === undefined) throw new TypeError("forged Auths recovery reference");
    return projectExecution("resume", await this.#call("resume", undefined, undefined, value));
  }

  async verify(value: ProductionAuthority | ProductionReceipt | Uint8Array): Promise<ProductionVerificationResult> {
    const bytes = value instanceof Uint8Array
      ? value
      : value.kind === "authority"
        ? readAuthority(value)
        : readReceipt(value);
    const projection = await this.#call("verify", bytes);
    if (projection.kind === "verified") {
      return Object.freeze({
        kind: "verified" as const,
        step: "verify" as const,
        ...(projection.value === null ? {} : { value: decodeBase64Url(projection.value) }),
      });
    }
    if (projection.kind === "rejected") {
      return Object.freeze({ kind: "rejected" as const, step: "verify" as const, code: requiredCode(projection), retry: "never" as const });
    }
    if (projection.kind === "indeterminate") return projectIndeterminate("verify", projection);
    throw new TypeError("native response outcome does not match verify");
  }

  async #call(
    step: ProductStep,
    body?: Uint8Array,
    authority?: Uint8Array,
    recoveryReference?: string,
  ): Promise<NativeProjection> {
    const engine = await productionEngine();
    const requestBody = engine.encodeProductionRequestV1({
      verb: step,
      profile: this.#profile.id,
      identity: this.#identity,
      ...(authority === undefined ? {} : { authority }),
      ...(body === undefined ? {} : { body }),
      ...(recoveryReference === undefined ? {} : { recoveryReference }),
    });
    const path = endpointPath(step, this.#profile.id);
    let response: ProductionTransportResponse;
    try {
      response = await this.#transport.send(Object.freeze({
        url: new URL(path, this.#endpoint),
        body: requestBody,
        contentType: CONTENT_TYPE,
        timeoutMs: this.#timeoutMs,
      }));
    } catch {
      return Object.freeze({
        contractVersion: 1,
        kind: "indeterminate",
        code: "core.runtime-unavailable",
        retry: "backoff",
        recoveryReference: null,
        value: null,
        receipt: null,
      });
    }
    if (response.status < 200 || response.status >= 300 || normalizeContentType(response.contentType) !== CONTENT_TYPE) {
      return Object.freeze({
        contractVersion: 1,
        kind: "indeterminate",
        code: "core.malformed-input",
        retry: "backoff",
        recoveryReference: null,
        value: null,
        receipt: null,
      });
    }
    if (!(response.body instanceof Uint8Array) || response.body.length === 0 || response.body.length > MAX_RESPONSE_BYTES) {
      throw new TypeError("Auths production response is outside bounds");
    }
    return parseProjection(engine.decodeProductionResponseV1(response.body));
  }
}

export function createProductionAuths(options: ProductionAuthsOptions): ProductionAuths {
  return new ProductionAuthsClient(options);
}

const fetchProductionTransport: ProductionTransport = Object.freeze({
  async send(request: ProductionTransportRequest): Promise<ProductionTransportResponse> {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), request.timeoutMs);
    try {
      const response = await fetch(request.url, {
        method: "POST",
        body: request.body as BodyInit,
        headers: Object.freeze({ "content-type": request.contentType, accept: request.contentType }),
        redirect: "error",
        credentials: "omit",
        signal: controller.signal,
      });
      const declared = Number(response.headers.get("content-length"));
      if (Number.isFinite(declared) && declared > MAX_RESPONSE_BYTES) {
        throw new TypeError("Auths production response is outside bounds");
      }
      const body = await readBoundedBody(response);
      return Object.freeze({
        status: response.status,
        contentType: response.headers.get("content-type") ?? "",
        body,
      });
    } finally {
      clearTimeout(timeout);
    }
  },
});

async function productionEngine(): Promise<NativeProductionEngine> {
  const engine = await loadPackagedWorkflowEngine();
  const native = engine as unknown as NativeProductionEngine;
  if (native.productionClientContractVersionV1() !== 1) {
    throw new TypeError("Auths production client contract mismatch");
  }
  return native;
}

async function encodeDelegationInput(subject: Uint8Array, attenuation: Uint8Array): Promise<Uint8Array> {
  if (!(subject instanceof Uint8Array) || subject.length === 0 || subject.length > 65_536
      || !(attenuation instanceof Uint8Array) || attenuation.length === 0 || attenuation.length > 65_536) {
    throw new TypeError("Auths delegation input is outside bounds");
  }
  return (await productionEngine()).encodeProductionDelegationV1(subject, attenuation);
}

function parseEndpoint(value: string | URL): URL {
  const endpoint = new URL(value);
  if (endpoint.protocol !== "https:" || endpoint.username !== "" || endpoint.password !== ""
      || endpoint.search !== "" || endpoint.hash !== "" || !["", "/"].includes(endpoint.pathname)) {
    throw new TypeError("Auths production endpoint must be an HTTPS origin");
  }
  return endpoint;
}

function isProductionProfile(value: ProductionProfile): boolean {
  return value !== null && typeof value === "object" && [
    "auths.opentofu.saved-plan-apply/1",
    "auths.postgresql.bounded-update/1",
    "auths.github.issue-address/1",
  ].includes(value.id);
}

function endpointPath(step: ProductStep, profile: ProductionProfile["id"]): string {
  if (step === "create") return "/v1/authority/create";
  if (step === "delegate") return "/v1/authority/delegate";
  if (step === "resume") return "/v1/workflows/resume";
  if (step === "verify") return "/v1/authority/verify";
  if (profile === "auths.opentofu.saved-plan-apply/1") return "/v1/profiles/opentofu/saved-plan-apply/execute";
  if (profile === "auths.postgresql.bounded-update/1") return "/v1/profiles/postgresql/bounded-update/execute";
  return "/v1/profiles/github/issue-address/execute";
}

function parseProjection(value: string): NativeProjection {
  const parsed: unknown = JSON.parse(value);
  if (parsed === null || typeof parsed !== "object") throw new TypeError("native response projection is invalid");
  const projection = parsed as Record<string, unknown>;
  if (projection.contractVersion !== 1
      || !["completed", "denied", "indeterminate", "recoverable", "verified", "rejected"].includes(String(projection.kind))
      || !["never", "backoff", "resume", "reconcile"].includes(String(projection.retry))) {
    throw new TypeError("native response projection is invalid");
  }
  return projection as unknown as NativeProjection;
}

function projectAuthorityFailure(step: "create" | "delegate", projection: NativeProjection): ProductionDenied | ProductionIndeterminate {
  if (projection.kind === "denied") {
    return Object.freeze({ kind: "denied", step, code: requiredCode(projection), retry: "never" });
  }
  if (projection.kind === "indeterminate") return projectIndeterminate(step, projection);
  throw new TypeError(`native response outcome does not match ${step}`);
}

function projectExecution(step: "execute" | "resume", projection: NativeProjection): ProductionExecutionResult {
  if (projection.kind === "completed") {
    if (projection.receipt === null) throw new TypeError("native response omitted receipt bytes");
    return Object.freeze({
      kind: "completed",
      step,
      ...(projection.value === null ? {} : { value: decodeBase64Url(projection.value) }),
      receipt: productionReceipt(decodeBase64Url(projection.receipt)),
    });
  }
  if (projection.kind === "denied") {
    return Object.freeze({ kind: "denied", step, code: requiredCode(projection), retry: "never" });
  }
  if (projection.kind === "indeterminate") return projectIndeterminate(step, projection);
  if (projection.kind === "recoverable" && projection.recoveryReference !== null) {
    return Object.freeze({
      kind: "recoverable",
      step,
      code: requiredCode(projection),
      retry: "resume",
      reference: productionRecoveryReference(projection.recoveryReference),
    });
  }
  throw new TypeError(`native response outcome does not match ${step}`);
}

function projectIndeterminate(step: ProductStep, projection: NativeProjection): ProductionIndeterminate {
  if (projection.retry !== "backoff" && projection.retry !== "reconcile") {
    throw new TypeError("native indeterminate result has invalid retry class");
  }
  return Object.freeze({ kind: "indeterminate", step, code: requiredCode(projection), retry: projection.retry });
}

function productionAuthority(bytes: Uint8Array): ProductionAuthority {
  if (bytes.length === 0) throw new TypeError("native response omitted authority bytes");
  return new ProductionAuthorityValue(bytes);
}

function productionReceipt(bytes: Uint8Array): ProductionReceipt {
  if (bytes.length === 0) throw new TypeError("native response omitted receipt bytes");
  return new ProductionReceiptValue(bytes);
}

function productionRecoveryReference(value: string): ProductionRecoveryReference {
  if (!/^[A-Za-z0-9_-]{43}$/.test(value)) {
    throw new TypeError("native response returned an invalid recovery reference");
  }
  return new ProductionRecoveryReferenceValue(value);
}

function requiredValue(projection: NativeProjection): Uint8Array {
  if (projection.value === null) throw new TypeError("native response omitted value bytes");
  return decodeBase64Url(projection.value);
}

function requiredCode(projection: NativeProjection): string {
  if (projection.code === null) throw new TypeError("native response omitted stable error code");
  return projection.code;
}

function readAuthority(authority: ProductionAuthority): Uint8Array {
  const value = authorityBytes.get(authority);
  if (value === undefined) throw new TypeError("forged Auths authority");
  return value.slice();
}

function readReceipt(receipt: ProductionReceipt): Uint8Array {
  const value = receiptBytes.get(receipt);
  if (value === undefined) throw new TypeError("forged Auths receipt");
  return value.slice();
}

function decodeBase64Url(value: string): Uint8Array {
  if (!/^[A-Za-z0-9_-]*$/.test(value)) throw new TypeError("native response encoding is invalid");
  const padding = "=".repeat((4 - value.length % 4) % 4);
  const binary = atob(value.replaceAll("-", "+").replaceAll("_", "/") + padding);
  return Uint8Array.from(binary, (character) => character.charCodeAt(0));
}

function normalizeContentType(value: string): string {
  return value.split(";", 1)[0]?.trim().toLowerCase() ?? "";
}

async function readBoundedBody(response: Response): Promise<Uint8Array> {
  if (response.body === null) return new Uint8Array();
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let length = 0;
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    length += value.length;
    if (length > MAX_RESPONSE_BYTES) {
      await reader.cancel();
      throw new TypeError("Auths production response is outside bounds");
    }
    chunks.push(value);
  }
  const body = new Uint8Array(length);
  let offset = 0;
  for (const chunk of chunks) {
    body.set(chunk, offset);
    offset += chunk.length;
  }
  return body;
}
